### ISSUE TITLE
Better support for extending NgxsDataEntityCollectionsRepository

### ISSUE THREAD

**Issue by bitflut** [ngxs-labs/data#360]
We recently switched from _ngrx_ to _nxgs_ with an app that heavily used _ngrx/data_. Right now we are trying to automate crud operations with our api in a similar way we did with _ngrx/data_. The route we've chosen is to **extend NgxsDataEntityCollectionsRepository** to include `getAll` `getOne` etc. for now.

Just to share what we are experiencing so far: If we are adding custom fields to the NgxsDataEntityCollectionsRepository reducer, we need to override `get snapshot()` and `getState()`. It'd be very nice, if there could be a third generic allowing to specify a type for the `NgxsEntityCollections`.

**Comment by splincode** [comment]
Give examples of expected behavior so that I can implement this

**Comment by bitflut** [comment]
I will put together a minimal example of what we're trying to achieve later today.

**Comment by bitflut** [comment]
Let's say we wanted additional properties in our EntityCollectionsReducer like `loading` and `loaded`:

```
export class MyCollectionRepository<V, K extends EntityIdType = EntityIdType>
    extends NgxsDataEntityCollectionsRepository<V, K> {

    @Computed()
    public get loading(): boolean {
        return this.snapshot.loading;
    }

    @Computed()
    public get loaded(): boolean {
        return this.snapshot.loaded;
    }

    @DataAction()
    public setLoading(@Payload('loading') loading: boolean): void {
        const state = this.getState();
        this.setEntitiesState({
            ...state,
            loading,
        });
    }

    @DataAction()
    public setLoaded(@Payload('loaded') loaded: boolean): void {
        const state = this.getState();
        this.setEntitiesState({
            ...state,
            loaded,
        });
    }
}
```

In order to get this working, there would be a few things we need to adjust. The most important is a new class type generic:

```
export class NgxsDataEntityCollectionsRepository<
    V,
    K extends number | string = EntityIdType,
    C extends NgxsEntityCollections<V, K> = NgxsEntityCollections<V, K>
>
```
Then we could simply use this in our Collection:
```
export class MyCollectionRepository<V, K extends EntityIdType = EntityIdType>
    extends NgxsDataEntityCollectionsRepository<V, K, MyCollection<V, K>>
```

There might be another caveat... for all the operations used in NgxsDataEntityCollectionsRepository, the rest of the state would need to be persisted. So `addEntityOne` would use this call to setEntitiesState:

```
        this.setEntitiesState({
            ...state,
            ids: [...state.ids, id],
            entities: { ...state.entities, [id]: entity },
        });
```

If this is hard to follow along, I can make an example repository with the overrides in place.

**Comment by splincode** [comment]
I understand you, yes, I think no problem will add this soon

**Comment by splincode** [comment]
```
### Extending NgxsDataEntityCollectionsRepository
```
move to Entity.md

**Comment by splincode** [comment]
released in `v3.0.0-rc.14`
